hail/python/test/hailtop/aiotools/test_copy.py

```diff
@@ -494,9 +500,8 @@ async def test_file_and_directory_error_with_slash_empty_file(router_filesystem)
             exp_dest = f'{dest_base}foo'
             await expect_file(fs, exp_dest, 'foo')
 
-
 @pytest.mark.asyncio
-async def test_file_and_directory_error_with_slash_non_empty_file(router_filesystem):
+async def test_file_and_directory_error_with_slash_non_empty_file_for_google_non_recursive(router_filesystem: Tuple[asyncio.Semaphore, AsyncFS, Dict[str, str]]):
```

Comment:
Can you explain this test to me and why you needed to split it into two?

Reply:
These tests check that we get errors when there is a file named `not-empty/` and a file named `not-empty/bar`. In the cloud, the existence of `not-empty/bar` implies the existence of a directory named `not-empty`. This should trigger a `FileAndDirectoryError`.

Google differs from Azure and S3 in how it lists files matching a prefix. Google considers the file named `gs://bucket/not-empty/` to be inside `gs://bucket` but not inside the "folder" named `gs://bucket/not-empty/`. Azure and AWS act differently. In Azure and AWS, the file named `/bucket/not-empty/` is considered to be inside the folder named `/bucket/not-empty/` (if such a folder exists).

As a result, on any cloud, you *should* get an error when recursively listing `gs://bucket` if it contains a file and directory with the same name. On AWS and Azure, you will not get an error when non-recursively listing `gs://bucket` because you never see both the file and the directory in the same list. If you list `/bucket` you only see the folder. If you list `/bucket/not-empty/` you only see the file. On GCP, if you list `/bucket` you see both `/bucket/not-empty/` the file and `/bucket/not-empty/` the folder.

In any cloud, if there exists a file-and-directory situation, an error will always occur when copying from the cloud to the local file system.

To accommodate the differing behavior of GCP, I moved the assertions that are only true for GCP into a pair of `...google_non_recursive` tests.